include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -53,6 +57,47 @@ __pattern_walk_n(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Function
     }
 }
 
+#if _ONEDPL_CPP20_RANGES_PRESENT
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_for_each
+//---------------------------------------------------------------------------------------------------------------------
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    auto __f_1 = 
+        [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+    oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), __f_1,
+                                                            oneapi::dpl::views::all(std::forward<_R>(__r)));
+
+    return std::ranges::borrowed_iterator_t<_R>(std::ranges::begin(__r) + std::ranges::size(__r));
```

Comment:
According to the standard, `std::ranges::for_each` returns `for_each_result<ranges::borrowed_iterator_t<R>, Fun>` rather than just `ranges::borrowed_iterator_t<R>`. Is there any reason to have such a difference?

Reply:
yes, there is a reason. 
Short answer  - F must be stateless for parallel execution => it doesnt make sense to return F with result.
More formal details here: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3179r1.html (see for_each)